File: manager-app/src/test/java/com/revature/CucumberTests/LoginSteps.java
package com.revature.CucumberTests;

import io.cucumber.java.Before;
import io.cucumber.java.en.*;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

import static org.junit.jupiter.api.Assertions.*;

public class LoginSteps {

    private final String BASE_URL = "http://localhost:5500";

    private WebDriverWait wait;

    public LoginSteps() {
        wait = new WebDriverWait(
                DriverManager.getDriver(),
                Duration.ofSeconds(5)
        );
    }

    @Given("I am on the manager login page")
    public void i_am_on_the_manager_login_page() {
        DriverManager.getDriver()
                .get(BASE_URL + "/login.html");
    }


    @When("I enter username {string} and password {string}")
    public void enterCredentials(String username, String password) {

        WebElement usernameInput = DriverManager.getDriver()
                .findElement(By.id("username"));

        WebElement passwordInput = DriverManager.getDriver()
                .findElement(By.id("password"));

        usernameInput.sendKeys(username);
        passwordInput.sendKeys(password);
    }

    @And("I click the login button")
    public void i_click_the_login_button() {
        DriverManager.getDriver()
                .findElement(By.xpath("//button[@type='submit']"))
                .click();
    }


    @Then("I should see a successful login message")
    public void i_should_see_a_successful_login_message() {
        wait.until(ExpectedConditions.textToBePresentInElementLocated(
                By.id("login-message"),
                "Login successful! Redirecting"
        ));

        assertTrue(
                DriverManager.getDriver()
                        .findElement(By.id("login-message"))
                        .getText()
                        .contains("Login successful! Redirecting")
        );
    }

    @Then("I should see an invalid credentials message")
    public void i_should_see_an_invalid_credentails_message() {
        wait.until(ExpectedConditions.textToBePresentInElementLocated(
                By.id("login-message"),
                "Invalid manager credentials"
        ));

        assertTrue(
                DriverManager.getDriver()
                        .findElement(By.id("login-message"))
                        .getText()
                        .contains("Invalid manager credentials")
        );
    }

    @Then("I should be redirected to the manager dashboard")
    public void i_should_be_redirected_to_the_manager_dashboard() {

        wait.until(ExpectedConditions.urlContains("manager"));

        assertTrue(
                DriverManager.getDriver()
                        .getCurrentUrl()
                        .contains("manager")
        );
    }


    @Then("I should remain on the login page")
    public void i_remain_on_the_login_page() {

        assertTrue(
                DriverManager.getDriver()
                        .getCurrentUrl()
                        .contains("login.html")
        );
    }
}

File: employee-app/features/steps/login_steps.py
from behave import given, when, then
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

@given('I am on the login page')
def step_on_login_page(context):
    context.driver.get(context.base_url + "/login")

@when('I enter username "{username}"')
def step_enter_username(context, username):
    context.driver.find_element(By.ID, "username").send_keys(username)

@when('I enter password "{password}"')
def step_enter_password(context, password):
    context.driver.find_element(By.ID, "password").send_keys(password)

@when('I click the login button')
def step_click_login(context):
    context.driver.find_element(By.XPATH, "//button[text()='Login']").click()

@then('I should see the title "{message}"')
def step_see_dashboard(context, message):
    heading = context.driver.find_element(By.TAG_NAME, "h1")
    assert message in heading.text, \
        f"Expected title '{message}', but got '{heading.text}'"


@then('I should see welcome message "{message}"')
def step_see_welcome(context, message):

    username = WebDriverWait(context.driver, 15).until(
        EC.text_to_be_present_in_element(
            (By.ID, "username-display"),
            "marco"
        )
    )

    assert username

@then('I should see an error message')
def step_see_error(context):
    # selenium is checking it too fast before the error message can show
    WebDriverWait(context.driver, 10).until(
        lambda d: d.find_element(By.ID, "login-message").text.strip() != ""
    )
